DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/mmhost/src/mmhost.cpp

```diff
@@ -0,0 +1,149 @@
+#include <iomanip>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/interfaces.hpp>
+#include <sycl/ext/oneapi/annotated_arg/annotated_ptr.hpp>
+#include <sycl/sycl.hpp>
+#include "exception_handler.hpp"
+
+using namespace sycl;
+using namespace ext::oneapi::experimental;
+using usm_buffer_location =
+    ext::intel::experimental::property::usm::buffer_location;
+
+constexpr int kBL1 = 1;
+constexpr int kBL2 = 2;
+constexpr int kBL3 = 3;
+
+struct PointerIP {
+  int *const a;
+  int *const b;
+  int *const c;
+  int n;
+
+  void operator()() const {
+    for (int i = 0; i < n; i++) {
+      c[i] = a[i] + b[i];
+    }
+  }
+};
+
+struct MMHostIP {
+  annotated_ptr<int, decltype(properties{buffer_location<kBL1>, awidth<32>,
+                                         dwidth<32>, latency<0>,
+                                         read_write_mode_readwrite, maxburst<4>,
+                                         wait_request_requested})>
+      a;
+  annotated_ptr<int, decltype(properties{buffer_location<kBL2>, awidth<15>,
+                                         dwidth<64>, latency<1>,
+                                         read_write_mode_read, maxburst<1>,
+                                         wait_request_not_requested})>
+      b;
+  annotated_ptr<int, decltype(properties{buffer_location<kBL3>, awidth<28>,
+                                         dwidth<16>, latency<16>,
+                                         read_write_mode_write, maxburst<1>,
+                                         wait_request_not_requested})>
+      c;
+
+  int n;
+
+  MMHostIP(int *a_, int *b_, int *c_, int n_) : a(a_), b(b_), c(c_), n(n_) {}
+
+  void operator()() const {
+    for (int i = 0; i < n; i++) {
+      c[i] = a[i] + b[i];
+    }
+  }
+};
+
+int main(void) {
+#if FPGA_SIMULATOR
+  auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+  auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+  auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+  try {
+    sycl::queue q(selector, fpga_tools::exception_handler,
+                  sycl::property::queue::enable_profiling{});
+
+    // Print out the device information.
+    std::cout << "Running on device: "
+              << q.get_device().get_info<sycl::info::device::name>() << "\n";
+
+    // Create and initialize the host arrays
+    constexpr int kN = 8;
+    std::cout << "elements in vector : " << kN << "\n";
+
+    auto host_array_A =
+        malloc_shared<int>(kN, q, property_list{usm_buffer_location(kBL1)});
+    assert(host_array_A);
+    auto host_array_B =
+        malloc_shared<int>(kN, q, property_list{usm_buffer_location(kBL2)});
+    assert(host_array_B);
+    auto host_array_C =
+        malloc_shared<int>(kN, q, property_list{usm_buffer_location(kBL3)});
+    assert(host_array_C);
+
+    auto pointer_array_A = malloc_shared<int>(kN, q);
```

Comment:
We might want to do these two kernels in separate files instead of one file. What do you think @whitepau , @ajaykumarkannan ?

If both kernels are in the same file the compiler treats them as two kernels in one system with the common set of global memories. Besides, [the HLS overview tutorial](https://github.com/intel-innersource/applications.fpga.oneapi.products.acl-source/tree/main/hls_example_designs/tutorials/interfaces/overview) nicely separated them out and I think that's a good model for a getting started tutorial which this is.

Here's what I think we should do to align with the excellent HLS tutorials:
1. Rename this tutorial as "interfaces/overview" or something similar and convert this cpp file into 2 cpp files - one with the non customized interfaces and one with the customized mmhost interfaces.
2. Add another cpp file for the register map based mmhost since that is missing here
3. In follow up changes we can either add cpp files for pipes tutorial - if we have an excellent pipes tutorial already we can link that here.
4. Add SYCL cpp file for the HLS tutorials pass_by_value example.

Hakim, get in touch with Paul to get his opinion on this. Let's sync with Ajay about this in our stand-up. I think if we do this we build a higher quality tutorial with better user experience. While we discuss this, try creating cpp files for (2) and (4)

Reply:
Synced up with Paul - Kevin is already working on an interfaces/overview equivalent tutorial. Hakim, you should then turn this into an equivalent of the [HLS mm_host specific tutorial](https://github.com/intel-innersource/applications.fpga.oneapi.products.acl-source/tree/main/hls_example_designs/tutorials/interfaces/pointer_mm_host).

Reply:
yes i would like multiple files. we should show these 3 examples:
1. naive (with unannotated poitners)
2. annotated to unique MM host interfaces (max throughput)
3. annotated to match DDR (see Readme_v3.zip)